### ISSUE TITLE
Unexpected velocity at end of drag gesture

### ISSUE THREAD

**Issue by stokesman** [pmndrs/use-gesture#332]
**Describe the bug**
Velocity seems to only be recalculated on movement events, making the last velocity likely to be inaccurate. This is notable mostly in cases where movement has stopped entirely.

**Sandbox**
https://codesandbox.io/s/use-gesture-drag-velocity-at-gesture-end-tu9kr?file=/src/index.js
Where it can be noted that dropping the draggable element after having ceased movement often still produces velocity and subsequent animation.

**Commentary**
In the sandbox linked above it might seem like a simple threshold test could be made and if the velocity is below it then further animation is avoided. That would work for probably most cases but would fail for high velocities that are possible after an abrupt halt to the movement (the demo button simulates that scenario). It could probably be worked around with a fairly simple timestamp test but this seems like a something the library could handle without detriment to other use cases.

**Information:**
 - Use Gesture version: 10
 - Device: macbook
 - OS: macOS
 - Browser: Brave

**Checklist:**
  - [ ] I've read the [documentation](https://use-gesture.netlify.app/).
  - [ ] If this is an issue with drag, I've tried setting `touch-action: none` to the draggable element.

**Comment by dbismut** [comment]
Hey @stokesman, first of all I'm happy that someone tried the vanilla version of this package. Let me know if there's something you'd like to be improved in the api.

Indeed this lib doesn't compute velocity for the last `pointerup` event (otherwise it would always be `0` since _most of the time_ the `pointerup` event coordinates are the same as the last `pointermove` event coordinates).

But I see the problem with abrupt movements: when the brutal halt occurs, the final `pointermove` event has some velocity,  and that velocity is retained instead of being set to `0`.

In my rapid tests, it looks like a throwing drag gesture generating deliberate inertia makes the `pointerup` event trigger about `~12ms` after the last `pointermove` event (measure on Firefox / Chrome desktop, Safari desktop gives me `~9ms`, Safari mobile also around `~8ms`). This means I could ignore computing velocity for `pointerup` events occurring within `20ms` or so from a `pointermove` event.

For `pointerup` events occurring _after_ `30ms` from a `pointermove` event, I would compte the velocity (this would result in the `velocity` being set to `0`).

Is this the heuristics that you had in mind? Of course I would prefer not relying on arbitrary figures that may vary overtime or change between hardware / browsers / OSes...

**Comment by stokesman** [comment]
I had imagined something like that for a workaround in userland. In the library, I think the better solution would be to decouple velocity calculation from the input events. I suppose a `requestAnimationFrame` loop during the gesture to handle such calculations may do the trick but I could be missing something. 

I appreciate the prompt reply though it's not an urgent matter on my end. Also it's great to see the vanilla version. I haven't used it enough to have feedback on the API.

**Comment by dbismut** [comment]
@stokesman the lib is supposed to be "just" a wrapper around events, it's not using `requestAnimationFrame` at all at the moment so it triggers user handlers in sync with the browser dispatching events. I'm not even sure how to reconciliate the event loop with a `requestAnimationFrame` loop, I guess that would end up complexifying the gesture engine by quite a lot 😅

I'll give a shot at setting an arbitrary delay before the last event in order to calculate the velocity. Let's see how this goes.

**Comment by dbismut** [comment]
For now relying on an arbitrary delay doesn't work consistently and results in some false positive / false negative... But it's an improvement over the previous logic, although less predictable.

**Comment by stokesman** [comment]
Hi @dbismut, I was curious and tried the suggestion I had made. It didn't change how user handlers were triggered and it worked nicely testing manually. It broke a few unit tests (velocity wasn't as predicted by input event timestamps used for tests) so I decided try something else.

I looked at how the Popmotion library had implemented velocity tracking and adapted that. It works quite well so I've put it up as #337. I'm not sure if it has any major advantage compared to what you've tried with the arbitrary delay, but at least it doesn't require such a thing.
